include/onnxruntime/core/graph/graph_nodes.h

```diff
@@ -48,19 +54,25 @@ class ValidNodes {
   }
 
   ConstReverseNodeIterator rbegin() const noexcept {
-    return {nodes_.crbegin(), nodes_.crend()};
+    return {nodes_.crbegin(), nodes_.crend(), filter_node_fn_};
   }
 
   ConstReverseNodeIterator rend() const noexcept {
-    return {nodes_.crend(), nodes_.crend()};
+    return {nodes_.crend(), nodes_.crend(), filter_node_fn_};
   }
 
-  MutableNodeIterator begin() noexcept {
-    return {nodes_.begin(), nodes_.end()};
+  // we only allow mutable access if the container is non-const.
+  // we need to templatize the functions for enable_if to work at this level, but mandate T2 being TNodesContainer
+  template <typename T2 = TNodesContainer>
+  typename std::enable_if<!std::is_const<T2>::value, MutableNodeIterator>::type begin() noexcept {
+    static_assert(std::is_same<T2, TNodesContainer>::value, "Explicit specialization is not allowed");
+    return MutableNodeIterator(nodes_.begin(), nodes_.end(), filter_node_fn_);
```

Comment:
Do you actually need this static_assert? Isn't T2=TNodesContainer already?

Reply:
It's defaulted to TNodesContainer. the static_assert enforces at compile time that nobody tries to override that.